dask_jobqueue/core.py

```diff
@@ -356,29 +354,29 @@ def stop_workers(self, workers):
 
     def stop_jobs(self, jobs):
         """ Stop a list of jobs"""
-        logger.debug("Stopping jobs: %s" % jobs)
+        logger.debug("Stopping jobs: %s", jobs)
         if jobs:
             jobs = list(jobs)
             self._call([self.cancel_command] + list(set(jobs)))
 
     def scale_up(self, n, **kwargs):
         """ Brings total worker count up to ``n`` """
-        logger.debug("Scaling up to %d workers." % n)
+        logger.debug("Scaling up to %d workers.", n)
         active_and_pending = sum([len(j) for j in self.running_jobs.values()])
         active_and_pending += self.worker_processes * len(self.pending_jobs)
-        logger.debug("Found %d active/pending workers." % active_and_pending)
+        logger.debug("Found %d active/pending workers.", active_and_pending)
         self.start_workers(n - active_and_pending)
 
     def scale_down(self, workers):
         ''' Close the workers with the given addresses '''
-        logger.debug("Scaling down. Workers: %s" % workers)
+        logger.debug("Scaling down. Workers: %s", workers)
         worker_states = []
         for w in workers:
```

Comment:
@mrocklin - I've been working on debugging the failures we're seeing and I may have stumbled onto a real bug. I'll lay out what I can understand.

- when we call `scale(0)`, we correctly end up in the scale down method
- however, `workers` is of type (`_asyncio.Future`) and is not a list of workers
- `workers.result()` yields an empty list. 

From here, I've tried a bunch of things including:

- removing the `loop.add_callback` calls in `scale`
- tweaking the remove/close_workers in `retire_workers`

But alas, I'm pretty lost. Any pointers would be appreciated.

Reply:
Yup, lost with reason.  That looks like a pretty clear bug.  I'm not sure how that ever would have worked.

For context.  `Scheduler.retire_workers` is a coroutine

```python
    @gen.coroutine
    def retire_workers(self, comm=None, workers=None, remove=True,
                       close_workers=False, **kwargs):
```

This means that it is meant to be called within *other* coroutines, with a yield statment

```python
@gen.coroutine
def f():
    yield scheduler.retire_workers(...)
```

This is true of any internal dask scheduler method that communicates with other things or does anything that might take any non-trivial amount of time.

In this way, coroutines are a bit viral, which in our case is bad, because we don't want `scale` to be a coroutine, because users use it, and users get confused by coroutines.

Another approach is to call coroutines with `add_callback` which says "run this whenever you have a moment, but don't lets deal with it now, because it's a bit messy".

So probably the thing to do here is in `distributed/deploy/cluster.py::Cluster.scale` we want to call `self.scheduler.workers_to_close` which is thankfully just a normal method and then we want to add `self.scheduler.retire_workers` as a callback with those workers to be run in just a moment, then we want to call `scale_down` as we do currently (which is also as a callback, just in case the cluster object implements it as a coroutine as well.

I can do this, but I'd also be very happy to guide someone else through it in the interests of spreading some of this knowledge around.

Reply:
To be clear the problem here is that we were calling it as a normal function

```python
def scale(...):
    ...

    to_close = self.scheduler.retire_workers(...)
```

*without a yield* within a function that *was not a coroutine*.  In this case it returns an opaque asyncio/torando Future object.  This is never a good thing to see and in our case almost always signifies a bug.

Reply:
@mrocklin - I'm happy to push this forward if you can provide some hand holding. https://github.com/jhamman/distributed/commit/ad0256006bc7d5018277da0d6c686224ffaaa8fe is a first attempt based on my reading of what is above. This does cause some tests to fail but I think it is worth sharing in its current form for initial feedback.